File: signal-lab/src/fade.cpp
#include "signal_lab/det_math.hpp"
#include "signal_lab/impairment.hpp"

#include <new>

namespace signal_lab
{
namespace
{

constexpr std::uint32_t max_active_fades = 8U;

class FadeImpairment final : public Impairment
{
public:
    explicit FadeImpairment(const FadeParams& params) noexcept : params_{params} {}

    [[nodiscard]] StageType type() const noexcept override
    {
        return StageType::fade;
    }

    [[nodiscard]] bool prepare(const StageContext& context, const char** error) noexcept override
    {
        if (params_.explicit_shape)
        {
            attack_ = det::seconds_to_frames(params_.attack_ms / 1e3, sample_rate_hz);
            hold_ = det::seconds_to_frames(params_.hold_ms / 1e3, sample_rate_hz);
            recovery_ = det::seconds_to_frames(params_.recovery_ms / 1e3, sample_rate_hz);
        }
        else
        {
            const std::uint64_t total = det::seconds_to_frames(params_.duration_ms / 1e3, sample_rate_hz);
            if (params_.rectangular)
            {
                attack_ = recovery_ = 0U;
                hold_ = total;
            }
            else
            {
                attack_ = recovery_ = total / 4U;
                hold_ = total - attack_ - recovery_;
            }
        }
        if (params_.rectangular)
        {
            hold_ += attack_ + recovery_;
            attack_ = recovery_ = 0U;
        }
        length_ = attack_ + hold_ + recovery_;
        if (length_ == 0U)
        {
            if (error != nullptr)
            {
                *error = "fade duration must span at least one sample";
            }
            return false;
        }
        depth_scale_ = -params_.depth_db * (det::ln10 / 20.0);
        resolve_window(params_.window, context.total_frames, window_start_, window_end_);
        total_frames_ = context.total_frames;
        rng_.reseed(context.seed, 8U, context.index);
        explicit_index_ = 0U;
        emitted_ = 0U;
        active_count_ = 0U;
        exhausted_ = false;
        position_ = static_cast<double>(window_start_);
        next_periodic_ = params_.has_first ? det::seconds_to_frames(params_.first_seconds, sample_rate_hz) : window_start_;
        stats_ = StageStats{};
        advance_pending();
        return true;
    }

    [[nodiscard]] std::size_t process(float* block, std::size_t frames, std::uint64_t first_input_frame, std::size_t) noexcept override
    {
        const std::uint64_t block_end = first_input_frame + frames;
        std::uint64_t cursor = first_input_frame;
        while (cursor < block_end)
        {
            // Retire at event boundaries, not once per input block. The bounded
            // slots represent simultaneous fades, including at a shared end/start.
            std::uint32_t kept = 0U;
            for (std::uint32_t slot = 0U; slot < active_count_; ++slot)
            {
                if (active_[slot] + length_ > cursor)
                {
                    active_[kept++] = active_[slot];
                }
            }
            active_count_ = kept;
            while (!exhausted_ && pending_ <= cursor)
            {
                if (active_count_ < max_active_fades)
                {
                    active_[active_count_++] = pending_;
                    stats_.events_applied++;
                }
                else
                {
                    stats_.events_dropped++;
                }
                advance_pending();
            }
            std::uint64_t segment_end = block_end;
            if (!exhausted_ && pending_ < segment_end)
            {
                segment_end = pending_;
            }
            for (std::uint32_t slot = 0U; slot < active_count_; ++slot)
            {
                const std::uint64_t end = active_[slot] + length_;
                if (end < segment_end)
                {
                    segment_end = end;
                }
            }
            // Keep the same chronological multiplication order for overlapping fades.
            for (std::uint32_t slot = 0U; slot < active_count_; ++slot)
            {
                const std::uint64_t start = active_[slot];
                for (std::uint64_t absolute = cursor; absolute < segment_end; ++absolute)
                {
                    const std::uint64_t u = absolute - start;
                    double depth = 1.0;
                    if (u < attack_)
                    {
                        depth = 0.5 * (1.0 - det::cos(det::pi * static_cast<double>(u) / static_cast<double>(attack_)));
                    }
                    else if (u >= attack_ + hold_)
                    {
                        const double phase = static_cast<double>(u - attack_ - hold_) / static_cast<double>(recovery_);
                        depth = 0.5 * (1.0 + det::cos(det::pi * phase));
                    }
                    const double gain = det::exp(depth_scale_ * depth);
                    float& sample = block[absolute - first_input_frame];
                    sample = static_cast<float>(static_cast<double>(sample) * gain);
                    stats_.component_frames++;
                    if (gain < minimum_gain_)
                    {
                        minimum_gain_ = gain;
                    }
                }
            }
            cursor = segment_end;
        }
        return frames;
    }

    [[nodiscard]] const StageStats& stats() const noexcept override
    {
        return stats_;
    }

private:
    void advance_pending() noexcept
    {
        if (params_.schedule == FadeSchedule::explicit_starts)
        {
            while (explicit_index_ < params_.start_count)
            {
                const std::uint64_t frame = det::seconds_to_frames(params_.starts_seconds[explicit_index_++], sample_rate_hz);
                if (total_frames_ == 0U || frame < total_frames_)
                {
                    pending_ = frame;
                    stats_.events_scheduled++;
                    return;
                }
            }
            exhausted_ = true;
            return;
        }
        if (params_.schedule == FadeSchedule::poisson)
        {
            position_ += rng_.exponential(static_cast<double>(sample_rate_hz) / params_.rate_per_sec);
            const double rounded = det::round_half_even(position_);
            if (rounded >= static_cast<double>(window_end_) || rounded < 0.0)
            {
                exhausted_ = true;
                return;
            }
            pending_ = static_cast<std::uint64_t>(rounded);
            stats_.events_scheduled++;
            return;
        }
        const std::uint64_t period = det::seconds_to_frames(params_.period_seconds, sample_rate_hz);
        if (period == 0U || next_periodic_ >= window_end_ || (params_.has_count && emitted_ >= params_.count) ||
            (total_frames_ != 0U && next_periodic_ >= total_frames_))
        {
            exhausted_ = true;
            return;
        }
        pending_ = next_periodic_;
        next_periodic_ += period;
        emitted_++;
        stats_.events_scheduled++;
    }

    FadeParams params_;
    std::uint64_t attack_{};
    std::uint64_t hold_{};
    std::uint64_t recovery_{};
    std::uint64_t length_{};
    double depth_scale_{};
    std::uint64_t window_start_{};
    std::uint64_t window_end_{};
    std::uint64_t total_frames_{};
    det::Pcg32 rng_{};
    double position_{};
    std::uint64_t next_periodic_{};
    std::uint32_t explicit_index_{};
    std::uint32_t emitted_{};
    std::uint64_t pending_{};
    bool exhausted_{};
    std::uint64_t active_[max_active_fades]{};
    std::uint32_t active_count_{};
    double minimum_gain_{1.0};
    StageStats stats_{};
};

static_assert(alignof(FadeImpairment) <= impairment_alignment);

} // namespace

std::size_t fade_size() noexcept
{
    return align_storage(sizeof(FadeImpairment));
}

Impairment* construct_fade(const FadeParams& params, void* storage, std::size_t storage_bytes) noexcept
{
    if (storage == nullptr || storage_bytes < fade_size())
    {
        return nullptr;
    }
    return new (storage) FadeImpairment(params);
}

} // namespace signal_lab

File: host/signal_lab/fade.py
"""Deterministic multiplicative amplitude fades (applied to the desired signal only).

Shape "raised_cosine": raised-cosine attack and recovery around a full-depth
hold (pcm_stress faded_signal); with only duration_ms given, attack =
recovery = duration/4 and hold = duration/2.  Shape "rectangular": full depth
for the whole duration with instantaneous edges.  Overlapping fades multiply
(depths add in dB).  Schedules: explicit starts_seconds, periodic
(first_seconds + period_seconds [+ count]) or Poisson (rate_per_sec, rng
family 8).
"""

import numpy as np

from . import FS
from .rng import FAMILY_FADE_SCHEDULE, generator
from .stream import Impairment, integer, number, seconds_to_frames

MAX_EVENTS = 100000


class Fade(Impairment):
    """params: depth_db, duration_ms | (attack_ms, hold_ms, recovery_ms), shape ('raised_cosine'|'rectangular'),
    starts_seconds [..] | (first_seconds, period_seconds[, count]) | rate_per_sec, start_seconds/duration_seconds window."""

    type_name = "fade"

    def prepare(self, reference_rms, total_frames, seed, index):
        super().prepare(reference_rms, total_frames, seed, index)
        p = self.params
        self.depth_db = number(p["depth_db"], "fade.depth_db", 0, 300)
        self.shape = p.get("shape", "raised_cosine")
        if self.shape not in ("raised_cosine", "rectangular"):
            raise ValueError("fade.shape must be raised_cosine or rectangular")
        if "attack_ms" in p or "hold_ms" in p or "recovery_ms" in p:
            self.attack = seconds_to_frames(number(p.get("attack_ms", 0), "fade.attack_ms", 0) / 1e3, "fade.attack_ms")
            self.hold = seconds_to_frames(number(p.get("hold_ms", 0), "fade.hold_ms", 0) / 1e3, "fade.hold_ms")
            self.recovery = seconds_to_frames(number(p.get("recovery_ms", 0), "fade.recovery_ms", 0) / 1e3, "fade.recovery_ms")
        else:
            total = seconds_to_frames(number(p["duration_ms"], "fade.duration_ms", 0.1) / 1e3, "fade.duration_ms", True)
            if self.shape == "rectangular":
                self.attack, self.recovery, self.hold = 0, 0, total
            else:
                self.attack = self.recovery = total // 4
                self.hold = total - self.attack - self.recovery
        if self.shape == "rectangular":
            self.hold += self.attack + self.recovery
            self.attack = self.recovery = 0
        self.length = self.attack + self.hold + self.recovery
        if self.length <= 0:
            raise ValueError("fade duration must span at least one sample")
        window_start = seconds_to_frames(p.get("start_seconds", 0.0), "fade.start_seconds")
        duration = p.get("duration_seconds")
        window_end = total_frames if duration is None else min(total_frames, window_start + seconds_to_frames(duration, "fade.duration_seconds", True))
        if "starts_seconds" in p:
            starts = sorted(seconds_to_frames(s, "fade.starts_seconds[]") for s in p["starts_seconds"])
            self.schedule = "explicit"
        elif "rate_per_sec" in p:
            rate = number(p["rate_per_sec"], "fade.rate_per_sec", 1e-6, 1000)
            rng = generator(seed, FAMILY_FADE_SCHEDULE, index)
            starts, position = [], float(window_start)
            while True:
                position += float(rng.exponential(FS / rate))
                frame = int(round(position))
                if frame >= window_end:
                    break
                starts.append(frame)
                if len(starts) > MAX_EVENTS:
                    raise ValueError("fade: too many events")
            self.schedule = "poisson"
        else:
            first = seconds_to_frames(p.get("first_seconds", window_start / FS), "fade.first_seconds")
            period = seconds_to_frames(p["period_seconds"], "fade.period_seconds", True)
            starts = list(range(first, window_end, period))
            if "count" in p:
                starts = starts[:integer(p["count"], "fade.count", 1, 2 ** 32 - 1)]
            self.schedule = "periodic"
        self.starts = [s for s in starts if s < total_frames]
        if len(self.starts) > MAX_EVENTS:
            raise ValueError("fade: too many events")
        self.cursor = 0
        self.faded_frames = 0
        self.minimum_gain = 1.0

    def gain(self, first_frame, count):
        gain = np.ones(count)
        absolute = np.arange(first_frame, first_frame + count)
        while self.cursor < len(self.starts) and self.starts[self.cursor] + self.length <= first_frame:
            self.cursor += 1
        touched = np.zeros(count, dtype=bool)
        for event in range(self.cursor, len(self.starts)):
            start = self.starts[event]
            if start >= first_frame + count:
                break
            t = absolute - start
            selected = (t >= 0) & (t < self.length)
            if not np.any(selected):
                continue
            u = t[selected].astype(np.float64)
            depth = np.ones(len(u))
            if self.attack:
                attack = u < self.attack
                depth[attack] = 0.5 * (1 - np.cos(np.pi * u[attack] / self.attack))
            if self.recovery:
                recovering = u >= self.attack + self.hold
                phase = (u[recovering] - self.attack - self.hold) / self.recovery
                depth[recovering] = 0.5 * (1 + np.cos(np.pi * phase))
            gain[selected] *= 10.0 ** (-self.depth_db * depth / 20.0)
            touched |= selected
        self.faded_frames += int(np.count_nonzero(touched))
        if np.any(touched):
            self.minimum_gain = min(self.minimum_gain, float(np.min(gain[touched])))
        return gain

    def process(self, block, first_frame):
        return block * self.gain(first_frame, len(block))

    def resolved(self):
        return {"type": self.type_name, "depth_db": self.depth_db, "shape": self.shape, "schedule": self.schedule,
                "attack_frames": self.attack, "hold_frames": self.hold, "recovery_frames": self.recovery,
                "event_frames": self.length, "event_count": len(self.starts),
                "convention": "multiplicative gain 10^(-depth_db*d(t)/20) on the desired signal only; raised-cosine d(t) over attack/recovery, 1 during hold; overlapping fades multiply"}

    def stats(self):
        return {"faded_frames": self.faded_frames, "faded_seconds": self.faded_frames / FS,
                "minimum_gain_db": 20 * np.log10(self.minimum_gain) if self.minimum_gain > 0 else None}

    def events(self):
        return [{"index": i, "start_frame": s, "start_seconds": s / FS, "end_frame": s + self.length} for i, s in enumerate(self.starts)]
